src/__internal__/focus-trap/focus-trap.component.js

```diff
@@ -0,0 +1,118 @@
+  autoFocus = true,
+  focusFirstElement,
+  bespokeTrap,
+}) => {
+  const ref = useRef();
+  const firstOpen = useRef(true);
+  const [focusableElements, setFocusableElements] = useState();
+  const [firstElement, setFirstElement] = useState();
+  const [lastElement, setLastElement] = useState();
+
+  function setElementFocus(element) {
+    if (typeof element === "function") {
+      element();
+    } else {
+      element.focus();
+    }
+  }
+
+  const hasNewInputs = useCallback(
+    (candidate) => {
+      if (!focusableElements || candidate.length !== focusableElements.length) {
+        return true;
+      }
+
+      return Array.from(candidate).some((el, i) => el !== focusableElements[i]);
+    },
+    [focusableElements]
+  );
+
+  useLayoutEffect(() => {
+    const elements = Array.from(
+      ref.current.querySelectorAll(defaultFocusableSelectors)
+    ).filter((el) => Number(el.tabIndex) !== -1);
+
+    if (hasNewInputs(elements)) {
+      setFocusableElements(Array.from(elements));
+      setFirstElement(elements[0]);
+      setLastElement(elements[elements.length - 1]);
+    }
+  }, [children, hasNewInputs]);
+
+  useEffect(() => {
+    if (autoFocus && firstOpen.current && (focusFirstElement || firstElement)) {
+      setElementFocus(focusFirstElement || firstElement);
+      firstOpen.current = false;
+    }
+  }, [autoFocus, firstElement, focusFirstElement]);
+
+  useEffect(() => {
+    const trapFn = (ev) => {
+      if (bespokeTrap) {
+        bespokeTrap(ev, firstElement, lastElement);
+        return;
+      }
+
+      if (ev.key === "Tab") {
+        if (!focusableElements.length) {
+          /* Block the trap */
+          ev.preventDefault();
+        } else if (ev.shiftKey) {
+          /* shift + tab */
+          if (document.activeElement === firstElement) {
+            lastElement.focus();
+            ev.preventDefault();
+          } else {
+            const currentIndex = focusableElements.findIndex(
+              (item) => item === document.activeElement
+            );
+
+            setElementFocus(focusableElements[currentIndex - 1]);
+            ev.preventDefault();
+          }
+        } else if (document.activeElement === lastElement) {
+          firstElement.focus();
+          ev.preventDefault();
+        } else {
+          const currentIndex = focusableElements.findIndex(
+            (item) => item === document.activeElement
+          );
+
+          setElementFocus(focusableElements[currentIndex + 1]);
+          ev.preventDefault();
+        }
+      }
+    };
+
+    document.addEventListener("keydown", trapFn);
+
+    return function cleanup() {
+      document.removeEventListener("keydown", trapFn);
+    };
+  }, [firstElement, lastElement, focusableElements, bespokeTrap]);
+
+  return <div ref={ref}>{children}</div>;
+};
+
+FocusTrap.propTypes = {
+  children: PropTypes.node.isRequired,
+  autoFocus: PropTypes.bool,
+  focusFirstElement: PropTypes.func,
+  bespokeTrap: PropTypes.func,
```

Comment:
I'd write prop descriptions here, as I'm not really sure what `bespokeTrap` does unless I check out the `FocusTrap` code